substrate/client/network/sync/src/strategy.rs

```diff
@@ -264,114 +350,145 @@ where
-		match self {
-			SyncingStrategy::WarpSyncStrategy(_) => {},
-			SyncingStrategy::StateSyncStrategy(_) => {},
-			SyncingStrategy::ChainSyncStrategy(strategy) =>
-				strategy.on_block_finalized(hash, number),
+		// Only `ChainSync` is interested in block finalization notifications.
+		if let Some(ref mut chain_sync) = self.chain_sync {
+			chain_sync.on_block_finalized(hash, number);
 		}
 	}
 
 	/// Inform sync about a new best imported block.
 	pub fn update_chain_info(&mut self, best_hash: &B::Hash, best_number: NumberFor<B>) {
-		match self {
-			SyncingStrategy::WarpSyncStrategy(_) => {},
-			SyncingStrategy::StateSyncStrategy(_) => {},
-			SyncingStrategy::ChainSyncStrategy(strategy) =>
-				strategy.update_chain_info(best_hash, best_number),
+		// This is relevant to `ChainSync` only.
+		if let Some(ref mut chain_sync) = self.chain_sync {
+			chain_sync.update_chain_info(best_hash, best_number);
 		}
 	}
 
 	// Are we in major sync mode?
 	pub fn is_major_syncing(&self) -> bool {
-		match self {
-			SyncingStrategy::WarpSyncStrategy(_) => true,
-			SyncingStrategy::StateSyncStrategy(_) => true,
-			SyncingStrategy::ChainSyncStrategy(strategy) =>
-				strategy.status().state.is_major_syncing(),
-		}
+		self.warp.is_some() ||
+			self.state.is_some() ||
+			match self.chain_sync {
+				Some(ref s) => s.status().state.is_major_syncing(),
+				None => unreachable!("At least one syncing startegy is active; qed"),
+			}
 	}
 
 	/// Get the number of peers known to the syncing strategy.
 	pub fn num_peers(&self) -> usize {
-		match self {
-			SyncingStrategy::WarpSyncStrategy(strategy) => strategy.num_peers(),
-			SyncingStrategy::StateSyncStrategy(strategy) => strategy.num_peers(),
-			SyncingStrategy::ChainSyncStrategy(strategy) => strategy.num_peers(),
-		}
+		self.peer_best_blocks.len()
 	}
 
 	/// Returns the current sync status.
 	pub fn status(&self) -> SyncStatus<B> {
-		match self {
-			SyncingStrategy::WarpSyncStrategy(strategy) => strategy.status(),
-			SyncingStrategy::StateSyncStrategy(strategy) => strategy.status(),
-			SyncingStrategy::ChainSyncStrategy(strategy) => strategy.status(),
+		// This function presumes that startegies are executed serially and must be refactored
+		// once we have parallel strategies.
+		if let Some(ref warp) = self.warp {
+			warp.status()
+		} else if let Some(ref state) = self.state {
+			state.status()
+		} else if let Some(ref chain_sync) = self.chain_sync {
+			chain_sync.status()
+		} else {
+			unreachable!("At least one syncing startegy is always active; qed")
 		}
 	}
 
 	/// Get the total number of downloaded blocks.
 	pub fn num_downloaded_blocks(&self) -> usize {
-		match self {
-			SyncingStrategy::WarpSyncStrategy(_) => 0,
-			SyncingStrategy::StateSyncStrategy(_) => 0,
-			SyncingStrategy::ChainSyncStrategy(strategy) => strategy.num_downloaded_blocks(),
+		if let Some(ref chain_sync) = self.chain_sync {
+			chain_sync.num_downloaded_blocks()
+		} else {
+			0
 		}
```

Comment:
```suggestion
        self.chain_sync.map_or(0, |chain_sync| chain_sync.num_downloaded_blocks())
```

Reply:
Unfortunately, `map()` consumes `self`, so this won't work. The closest possible solution is probably converting option into iterator, but it's quite ugly and would be even uglier if you need not only to call a function, but return a result.

Reply:
what about `self.chain_sync.as_ref().map()`?